### DISCUSSION TITLE
Allow 'required' keyword in positional syntax for record structs

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/csharplang#6760] [Language Ideas]
Correct me if I'm wrong, but it seems to me that with `required` keyword introduced in C# 11 we finally have a way to stop users of our `struct` types from blindly initializing them using default constructor:
```csharp
public class Program
{
    readonly record struct AgeNotRequired(int Age);

    readonly record struct AgeRequired
    {
        public required int Age { get; init; }
    }

    public static void Main(string[] args)
    {
        var ageNotRequired = new AgeNotRequired(); // Compiles without errors
        var ageRequired = new AgeRequired(); // Error: Required member 'Program.AgeRequired.Age' must be set in the object initializer
    }
}
```

This is nice, however it would be even nicer to just write:

```csharp
public class Program
{
    readonly record struct AgeNotRequired(int Age);

    readonly record struct AgeRequired(required int Age);

    public static void Main(string[] args)
    {
        var ageNotRequired = new AgeNotRequired(); // Compiles without errors
        var ageRequired = new AgeRequired(); // Error: Required member 'Program.AgeRequired.Age' must be set...
    }
}
```
This only seems relevant for struct types, because for regular records you cannot skip uninitialized positional properties.

**Comment by NeilMacMullen** [comment]
The problem is actually worse in practice.  It's possible to write 

```
readonly record struct MyRecord(string Name,string Description);
var r = new MyRecord { 
     Name= "foo"
     /* Description not supplied ! */ 
};
```

Since in these cases I almost always want my records to be immutable AND complete, it's frustrating that there is this loophole.  It would be nice if we could add "required" next to "readonly" or even better some new keyword that would indicate the combination of immutabilty and completeness.

**Reply by RikkiGibson** [reply]
Reminds me of the following:

```cs
// making an array, nothing to see here :)
var arr = new ImmutableArray<int>() { 1, 2, 3 };
```

Perhaps better because at least it throws an exception at runtime, but man, it's not great that nothing along the way gives you a heads up here.

**Reply by Joe4evr** [reply]
@RikkiGibson Sounds like a job for https://github.com/dotnet/roslyn-analyzers. 🙃
